pkg/pool-weighted/contracts/managed/ManagedPoolCircuitBreakerLib.sol

```diff
@@ -0,0 +1,248 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/WordCodec.sol";
+
+import "../lib/ValueCompression.sol";
+
+/**
+ * @title Managed Pool Circuit Breaker Library
+ * @notice Library for storing and manipulating state related to circuit breakers in ManagedPool.
+ * @dev BPT price = supply / balance / weight. The lower and upper bounds are percentages.
+ * For instance, a lower bound of 0.8 means the circuit breaker will trip if the token price
+ * drops more than 20% (relative to other tokens). An upper bound ratio of 2.0 means it will trip
+ * if the price more than doubles.
+ */
+library ManagedPoolCircuitBreakerLib {
+    using ValueCompression for uint256;
+    using FixedPoint for uint256;
+    using WordCodec for bytes32;
+
+    struct CircuitBreakerParams {
+        uint256 referenceBptPrice;
+        uint256 currentWeightFactor;
+        uint256 lowerBoundPercentage;
+        uint256 upperBoundPercentage;
+    }
+
+    // Store circuit breaker information per token
+    // [  24 bits |  24 bits |   16 bits   |   16 bits   |    64 bits    |  112 bits |
+    // [ UB ratio | LB ratio | upper bound | lower bound | weight factor | ref price |
+    // |MSB                                                                       LSB|
+    uint256 private constant _REFERENCE_BPT_PRICE_OFFSET = 0;
+    uint256 private constant _WEIGHT_FACTOR_OFFSET = _REFERENCE_BPT_PRICE_OFFSET + _BPT_PRICE_WIDTH;
+    uint256 private constant _LOWER_BOUND_PCT_OFFSET = _WEIGHT_FACTOR_OFFSET + _WEIGHT_FACTOR_WIDTH;
+    uint256 private constant _UPPER_BOUND_PCT_OFFSET = _LOWER_BOUND_PCT_OFFSET + _BOUND_PERCENTAGE_WIDTH;
+    uint256 private constant _LOWER_BOUND_RATIO_CACHE_OFFSET = _UPPER_BOUND_PCT_OFFSET + _BOUND_PERCENTAGE_WIDTH;
+    uint256 private constant _UPPER_BOUND_RATIO_CACHE_OFFSET = _LOWER_BOUND_RATIO_CACHE_OFFSET +
+        _BOUND_RATIO_CACHE_WIDTH;
```

Comment:
Well, it's a cache in the sense that it is stored in the _circuitBreakerState when it is originally set along with the weight factor. If nothing has changed (no ongoing weight change, or tokens added/removed), we can just use these values directly, instead of performing expensive exponentiation.

I suppose I could call everything stored for this purpose the "reference" value, like the bptPrice is, to avoid overloading the "cache" concept.